omnipaxos_core/src/sequence_paxos/mod.rs

```diff
@@ -213,6 +214,43 @@ where
         }
     }
 
+    /// Detects if a Prepare, AcceptStopSign, or PrepareReq message has been sent but not received a reply.
+    /// If so resends them.
+    pub(crate) fn resend_message_timeout(&mut self) {
+        // Resend Prepare
+        if self.state.0 == Role::Leader {
+            let unpromised_peers = self.leader_state.get_unpromised_peers();
+            for peer in unpromised_peers {
+                self.send_prepare(peer);
+            }
+        }
+
+        // Resend AcceptStopSign
+        if self.state == (Role::Leader, Phase::Accept) {
+            // TODO: This is slow. Get stopsign from cache instead.
+            match self.internal_storage.get_stopsign() {
+                Some(ss) if !ss.decided => {
+                    for follower in self.leader_state.get_promised_followers() {
+                        if !self.leader_state.get_accepted_stopsign(follower) {
+                            // TODO: subtract 1 from seq_num to show duplicated?
+                            self.send_accept_stopsign(follower, ss.stopsign.clone());
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+
+        // Resend PrepareReq
+        if self.state == (Role::Follower, Phase::Recover) {
```

Comment:
Let's pattern match directly on `self.state` something like this:

```
match self.state {
       (Role::Leader, phase) => {
                  // resend prepare
                  // resend stopsign if there is a pending stopsign and phase == Phase::accept. 
       }, 
      (Role::Follower, Phase::Recover) => ...
}
```

Reply:
I'm not sure I agree. It is more readable because you first only match on the `Role` (which is in fact either `Leader` or `Follower`) while in your original code, we have three ifs even though some of them should be if/else.

I also don't see a problem with matching `() => _`. You can also get rid of one of them by doing `if let Some(...)` when checking the `StopSign`.

Two side notes:
1. Do we need to `clone` the promise? Ballots should be `copy`, and if we really need to clone it, we should ideally do it inside the `match` or `if`.
2. I think the servers are initialized to be in the Prepare phase, so it could be that we're in the Prepare phase without having a promise which can cause unnecessary logging.